// @grpc/grpc-js/src/filter.ts


import { StatusObject, WriteObject } from './call-interface';
import { Metadata } from './metadata';


export interface Filter {
  sendMetadata(metadata: Promise<Metadata>): Promise<Metadata>;

  receiveMetadata(metadata: Metadata): Metadata;

  sendMessage(message: Promise<WriteObject>): Promise<WriteObject>;

  receiveMessage(message: Promise<Buffer>): Promise<Buffer>;

  receiveTrailers(status: StatusObject): StatusObject;
}

export abstract class BaseFilter implements Filter {
  async sendMetadata(metadata: Promise<Metadata>): Promise<Metadata> {
    return metadata;
  }

  receiveMetadata(metadata: Metadata): Metadata {
    return metadata;
  }

  async sendMessage(message: Promise<WriteObject>): Promise<WriteObject> {
    return message;
  }

  async receiveMessage(message: Promise<Buffer>): Promise<Buffer> {
    return message;
  }

  receiveTrailers(status: StatusObject): StatusObject {
    return status;
  }
}

export interface FilterFactory<T extends Filter> {
  createFilter(): T;
}
