// web-vitals/src/attribution/onFID.ts


import {getLoadState} from '../lib/getLoadState.js';
import {getSelector} from '../lib/getSelector.js';
import {onFID as unattributedOnFID} from '../onFID.js';
import {
  FIDAttribution,
  FIDMetric,
  FIDMetricWithAttribution,
  ReportOpts,
} from '../types.js';

const attributeFID = (metric: FIDMetric): FIDMetricWithAttribution => {
  const fidEntry = metric.entries[0];
  const attribution: FIDAttribution = {
    eventTarget: getSelector(fidEntry.target),
    eventType: fidEntry.name,
    eventTime: fidEntry.startTime,
    eventEntry: fidEntry,
    loadState: getLoadState(fidEntry.startTime),
  };

  
  const metricWithAttribution: FIDMetricWithAttribution = Object.assign(
    metric,
    {attribution},
  );
  return metricWithAttribution;
};


export const onFID = (
  onReport: (metric: FIDMetricWithAttribution) => void,
  opts?: ReportOpts,
) => {
  unattributedOnFID((metric: FIDMetric) => {
    const metricWithAttribution = attributeFID(metric);
    onReport(metricWithAttribution);
  }, opts);
};
