// web-vitals/src/attribution/onTTFB.ts


import {onTTFB as unattributedOnTTFB} from '../onTTFB.js';
import {
  TTFBMetric,
  TTFBMetricWithAttribution,
  ReportOpts,
  TTFBAttribution,
} from '../types.js';

const attributeTTFB = (metric: TTFBMetric): TTFBMetricWithAttribution => {
  
  let attribution: TTFBAttribution = {
    waitingDuration: 0,
    cacheDuration: 0,
    dnsDuration: 0,
    connectionDuration: 0,
    requestDuration: 0,
  };

  if (metric.entries.length) {
    const navigationEntry = metric.entries[0];
    const activationStart = navigationEntry.activationStart || 0;

    
    
    
    const waitEnd = Math.max(
      (navigationEntry.workerStart || navigationEntry.fetchStart) -
        activationStart,
      0,
    );
    const dnsStart = Math.max(
      navigationEntry.domainLookupStart - activationStart,
      0,
    );
    const connectStart = Math.max(
      navigationEntry.connectStart - activationStart,
      0,
    );
    const connectEnd = Math.max(
      navigationEntry.connectEnd - activationStart,
      0,
    );

    attribution = {
      waitingDuration: waitEnd,
      cacheDuration: dnsStart - waitEnd,
      
      
      dnsDuration: connectStart - dnsStart,
      connectionDuration: connectEnd - connectStart,
      
      
      
      
      requestDuration: metric.value - connectEnd,
      navigationEntry: navigationEntry,
    };
  }

  
  const metricWithAttribution: TTFBMetricWithAttribution = Object.assign(
    metric,
    {attribution},
  );
  return metricWithAttribution;
};


export const onTTFB = (
  onReport: (metric: TTFBMetricWithAttribution) => void,
  opts?: ReportOpts,
) => {
  unattributedOnTTFB((metric: TTFBMetric) => {
    const metricWithAttribution = attributeTTFB(metric);
    onReport(metricWithAttribution);
  }, opts);
};
